exclude default paths: change from command line parameter to language file flag

miguel-negrao opened the pull request:
Before this PR:
 - default paths were excluded from the list of directories to search for class files if sclang was started with -a command line option.
- The "Exclude default search paths" option on the interpreter options in scide would toggle the use of `sclang -a`.

with this PR:
 - An additional method is added to exclude the default paths excluded from the list of directories to search for. To turn on or off excluding the default paths the 'excludeDefaultPaths' key can be used in the YAML language file.
 - The reasoning behind this PR is that some language config files require the default include paths to be active while others require the default include paths not to be active, in order for the language to compile successfully, therefore it makes sense that that choice is made in the yaml file itself, so that it will work automatically. This way switching to a different language config file will also switch whether default paths are excluded or not.
- The behaviour of `sclang -a` is kept the same.
- The "Exclude default search paths" option on the interpreter options in scide, doesn't use `sclang -a` anymore, instead just mirrors the value of 'excludeDefaultPaths' key  in the YAML language file. Before, with this PR the default paths were excluded for ALL language config files, now the choice is done in each YAML language file.
- A getter and setter were added for excludeDefaultPaths in the LanguageConfig class library class.
- In PyrLexer.cpp changes were made such that when recompiling the library the in-memory version of the LanguageConfig will be used instead of the version in the YAML file. This allows testing language settings by manipulation through the language (via LanguageConfig.sc) before writing to the file.

An example config file that uses a Class Library other than the system one:

```
includePaths:
    -   /home/miguel/Development/SuperCollider/supercollider/SCClassLibrary
excludePaths:
    []
postInlineWarnings: false
excludeDefaultPaths: true
```

miguel-negrao commented:
Tests that fail actually compile and run sclang... not sure what is the problem, it just says "ERROR:
Process not running".

brianlheim commented:
Take a look at what qpm is doing. It may rely on `-a`. This would be my first comment about this PR anyway - "deprecated" does not mean "removed", we need to keep the flag for at least one version so users have time to adjust. In fact, I see no reason to remove it, since doing so requires clients to create their own yaml config file to get the same result.

miguel-negrao commented:
> Take a look at what qpm is doing. It may rely on -a. This would be my first comment about this PR anyway - "deprecated" does not mean "removed", we need to keep the flag for at least one version so users have time to adjust. In fact, I see no reason to remove it, since doing so requires clients to create their own yaml config file to get the same result.

I think that if you exclude the default paths with the command line flag then you need a language file anyway in order to include at least a class library, no ? If that is the case then a language file is always required when excluding default paths. Or am I misunderstanding your issue ?

Ok, about not removing the functionality via -a option for this version. will update PR.

brianlheim commented:
> I think that if you exclude the default paths with the command line flag then you need a language file anyway in order to include at least a class library, no ?

Good point; it's been a long time since I looked at this and I assumed there was a way to include paths from the CLI. There is a flag for that in scsynth but not sclang.

miguel-negrao commented:
Ok, updated PR, it is ready for review again. I've opted for in case **either** the -a option is invoked or the sclang conf file contains the excludeDefaultPaths key then default paths are excluded.

So, given that a language file is necessary if the -a is invoked, I think in the version after the version where this PR will be included the -a option should be removed altogether to simplify code and not have two ways of doing the same thing. If we agree that -a should be deprecated (and will be removed), is it enough to mention that in the documentation or is it necessary to also send a message about to stderr ?

brianlheim commented:
> Ok, updated PR, it is ready for review again. I've opted for in case either the -a option is invoked or the sclang conf file contains the excludeDefaultPaths key then default paths are excluded.

That seems reasonable to me.

> So, given that a language file is necessary if the -a is invoked, I think in the version after the version where this PR will be included the -a option should be removed altogether to simplify code and not have two ways of doing the same thing. If we agree that -a should be deprecated (and will be removed), is it enough to mention that in the documentation or is it necessary to also send a message about to stderr ?

For now, I don't agree that it should be deprecated. This code is relatively low in maintenance cost and we may in the future add the ability to include/exclude paths as command-line options. In that case, we would need to reimplement this logic.

brianlheim commented on lang/LangPrimSource/PyrPrimitive.cpp:
```diff
@@ -4221,6 +4227,7 @@ void initPrimitives()
 	definePrimitive(base, index++, "_LanguageConfig_writeConfigFile", prLanguageConfig_writeConfigFile, 2, 0);
 	definePrimitive(base, index++, "_LanguageConfig_getPostInlineWarnings", prLanguageConfig_getPostInlineWarnings, 1, 0);
 	definePrimitive(base, index++, "_LanguageConfig_setPostInlineWarnings", prLanguageConfig_setPostInlineWarnings, 2, 0);
+	definePrimitive(base, index++, "_LanguageConfig_getExcludeDefaultPaths", prLanguageConfig_getExcludeDefaultPaths, 1, 0);
```
there should be a corresponding setter

brianlheim commented on lang/LangSource/SC_LanguageConfig.hpp:
```diff
@@ -65,6 +66,11 @@ class SC_LanguageConfig
 	bool forEachIncludedDirectory(bool (*)(const Path&)) const;
 
+	bool getExcludeDefaultPaths() const;
+	void setExcludeDefaultPaths(bool value);
```
these can be defined inline, like `get/setPostInlineWarnings` below, to save space

brianlheim commented on lang/LangSource/SC_LanguageConfig.hpp:
```diff
@@ -84,6 +92,7 @@ class SC_LanguageConfig
 	DirVector mIncludedDirectories;
 	DirVector mExcludedDirectories;
 	DirVector mDefaultClassLibraryDirectories;
+	bool mExcludeDefaultPaths = false;
```
also, I much prefer this as a member variable than a global, thanks for that

brianlheim commented on lang/LangSource/SC_LanguageConfig.hpp:
```diff
@@ -74,7 +80,9 @@ class SC_LanguageConfig
-	static const void setConfigPath(const Path& p) { gConfigFile = p; }
+	static const void setConfigPath(const Path& p) {
+		gConfigFile = p;
+	}
```
why did this change? please change it back so we don't have commit noise

brianlheim commented on lang/LangSource/SC_LanguageConfig.hpp:
```diff
@@ -65,6 +66,11 @@ class SC_LanguageConfig
+	bool getExcludeDefaultPaths() const;
+	void setExcludeDefaultPaths(bool value);
+	void static processPathList(const char* nodeName, YAML::Node &doc, void (*func)(const Path&));
+	void static processBool(const char* nodeName, YAML::Node &doc, const std::function<void(bool)> &func);
```
these don't need to be declared here, they could be static methods in the implementation file. that keeps stuff out of the public interface here.

brianlheim replied:
this would also allow you to remove the `#include <yaml-cpp/yaml.h>` from this header

brianlheim replied:
also, style in this codebase is overwhelmingly `static void` rather than `void static`.

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -237,7 +258,7 @@ void SC_LanguageConfig::freeLibraryConfig()
 	if (gLanguageConfig) {
 		delete gLanguageConfig;
-		gLanguageConfig = 0;
+		gLanguageConfig = nullptr;
```
:+1:

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -111,48 +114,64 @@ bool SC_LanguageConfig::removeExcludedDirectory(const Path& path)
+				gLanguageConfig = new SC_LanguageConfig(b || standalone);
+			});
+			if (!gLanguageConfig)
+				gLanguageConfig = new SC_LanguageConfig(standalone);
```
this logic was fine the way it was, with SC_LanguageConfig being created at the top of the function. You could use `SC_LanguageConfig::setExcludeDefaultPaths` here.

miguel-negrao replied:
Adding the default paths is done in the constructor of SC_LanguageConfig. You cannot call the constructor at the top of the function because you still haven't parsed the excludeDefaultPaths flag from the yaml file, and therefore you can't pass it to the constructor. This is why the object is created only after trying to get the excludeDefaultPaths flag from the yaml file. After having tried different approaches I came to prefer maintaining the constructor as it is now, where it adds the default paths in the constructor, I think the code is simpler that way.

brianlheim replied:
you could also assign a boolean and call the constructor later. more to the point, the design of this class should change so that a setter function is feasible. AFAICT there's no good reason not to allow that aside from the current implementation being what it is.

miguel-negrao replied:
> more to the point, the design of this class should change so that a setter function is feasible. AFAICT there's no good reason not to allow that aside from the current implementation being what it is.

My thoughts on this:

* The value of mDefaultClassLibraryDirectories is only relevant after reading from the YAML file and before compiling the class library (in PyrLexer.cpp passOne() ).  I don't think it is ever used again after that point. 
* Therefore mDefaultClassLibraryDirectories will have to be set very soon after the SC_LanguageConfig is created.
* Setting mDefaultClassLibraryDirectories in SC_LanguageConfig::setExcludeDefaultPaths(bool value) is ok, but actually it will not serve any purpose after the language is compiled, although that is also not a big issue, so I can implement it this way if you really prefer it.
* Given than mDefaultClassLibraryDirectories must be set after/while reading from the YAML file, might as well make this obvious, by initializing mDefaultClassLibraryDirectories in the constructor of SC_LanguageConfig which is called at that point.
* I think both options have pros and cons, so I'm OK with either option.

miguel-negrao replied:
Also, there is already a setter function, it just doesn't affect mDefaultClassLibraryDirectories for the reasons given above.  Unrelated to this issue, I will implement the primitive for the setter to be able to set it from sclang. That is only usefull when saving back to YAML.

jamshark70 replied:
> Unrelated to this issue, I will implement the primitive for the setter to be able to set it from sclang. That is only usefull when saving back to YAML.

IMO that's reason enough to do it. If it isn't part of the LanguageConfig spec (hmm...), it should be: that all aspects of the configuration should be possible to set using language code, and written to disk.

IIRC the behavior is supposed to be:

- If you use sclang methods to alter the configuration and recompile the class library, the compiler should then use the in-memory (modified, new) image of the configuration immediately without re-reading the configuration from disk.
  - Use case: Test a configuration change in memory, without writing. If the change broke the environment, the original, working configuration is still on disk.

- If you modify the configuration outside of sclang (including IDE preferences), the only way to use the changes is to save them to disk and restart sclang. Recompiling the library doesn't work in that case.

So the setter would be useful also for recompiling the library.

(Again, confusion because nobody bothered to write down what is the expected behavior.)

miguel-negrao replied:
Ok, that is worth exploring. A slightly different approach could be:
* When recompiling the library the in-memory version is always used.
* On sclang startup it reads the language from disk.
* When changing the language config from the IDE, the IDE tells the language to re-read the configuration from disk, so as to pick up the changes when the language is recompiled. 
* Changes made to the config file manually or in other software will only be picked up after re-starting sclang.

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -111,48 +114,64 @@ bool SC_LanguageConfig::removeExcludedDirectory(const Path& path)
+void SC_LanguageConfig::processPathList(const char* nodeName, YAML::Node &doc, void (*func)(const Path&)) {
+	const YAML::Node & items = doc[ nodeName ];
+	if (items && items.IsSequence()) {
+		std::string emptyString;
```
you can remove this and write `item.as<std::string>("")` below, it's cleaner.

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -111,48 +114,64 @@ bool SC_LanguageConfig::removeExcludedDirectory(const Path& path)
+{
+	return mExcludeDefaultPaths;
+}
+void SC_LanguageConfig::processPathList(const char* nodeName, YAML::Node &doc, void (*func)(const Path&)) {
```
better use `std::function` here for consistency

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -111,48 +114,64 @@ bool SC_LanguageConfig::removeExcludedDirectory(const Path& path)
+			processBool(POST_INLINE_WARNINGS, doc, [](bool b){
+				gPostInlineWarnings = b;
+			});
+			processBool(EXCLUDE_DEFAULT_PATHS, doc, [&standalone](bool b){
```
copy by value is better for built-in integral types

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -42,13 +43,15 @@ static const char* EXCLUDE_PATHS = "excludePaths";
 
-SC_LanguageConfig::SC_LanguageConfig(bool optStandalone)
+SC_LanguageConfig::SC_LanguageConfig(bool excludeDefaultPaths)
+	:mExcludeDefaultPaths(excludeDefaultPaths)
```
it's more conventional to write this as:
```cpp
Object::Object():
    member1(),
    member2()
```

miguel-negrao commented:
@brianlheim  thanks for the review ! I will have a look at your comments soon.

miguel-negrao commented on editors/sc-ide/widgets/settings/sclang_page.cpp:
```diff
@@ -76,7 +77,6 @@ void SclangPage::load( Manager *s )
 
     ui->autoStart->setChecked( s->value("autoStart").toBool() );
     ui->runtimeDir->setText( s->value("runtimeDir").toString() );
-    ui->sclang_standalone_mode->setChecked(s->value("standalone").toBool());
```
Since this setting is now saved in the language file, I reasoned that it didn't make sense to also save it in the IDE preferences. The widget is checked or not based on the value saved in the lang config file.

miguel-negrao commented on editors/sc-ide/widgets/settings/sclang_page.cpp:
```diff
@@ -99,7 +99,6 @@ void SclangPage::store( Manager *s )
     s->setValue("autoStart", ui->autoStart->isChecked());
     s->setValue("runtimeDir", ui->runtimeDir->text());
     s->setValue("configFile", ui->activeConfigFileComboBox->currentText());
-    s->setValue("standalone", ui->sclang_standalone_mode->isChecked());
```
same as above

miguel-negrao commented:
> So, given that a language file is necessary if the -a is invoked, I think in the version after the version where this PR will be included the -a option should be removed altogether to simplify code and not have two ways of doing the same thing. If we agree that -a should be deprecated (and will be removed), is it enough to mention that in the documentation or is it necessary to also send a message about to stderr ?
> 
> For now, I don't agree that it should be deprecated. This code is relatively low in maintenance cost and we may in the future add the ability to include/exclude paths as command-line options. In that case, we would need to reimplement this logic.

Ok, we can discuss later if should be deprecated or not, for this PR we don't deprecate anything.

@brianlheim I've commented on a couple of your comments where I would like to discuss further. All other comments I agree with and will change the code based on them.

miguel-negrao commented on lang/LangSource/PyrLexer.cpp:
```diff
@@ -2111,7 +2110,9 @@ SCLANG_DLLEXPORT_C bool compileLibrary(bool standalone)
-	SC_LanguageConfig::readLibraryConfig(standalone);
+	if (!gLanguageConfig) {
+		SC_LanguageConfig::readLibraryConfig(standalone);
+	}
```
With this change we don't read the config file from disk and instead use the in memory version

miguel-negrao commented on lang/LangSource/SC_LanguageConfig.hpp:
```diff
@@ -84,6 +85,7 @@ class SC_LanguageConfig
 	DirVector mIncludedDirectories;
 	DirVector mExcludedDirectories;
 	DirVector mDefaultClassLibraryDirectories;
+	bool mExcludeDefaultPaths = true;
```
true at instatiation time in order to mirror fact that mDefaultClassLibraryDirectories vector is empty.

miguel-negrao commented on lang/LangSource/PyrLexer.cpp:
```diff
@@ -2099,7 +2099,6 @@ void shutdownLibrary()
 	gLangMutex.unlock();
 	deinitPrimitives();
 
-	SC_LanguageConfig::freeLibraryConfig();
```
On shutdownLibrary() library is not deleted in order to be able to use in-memory version on recompile.

miguel-negrao commented:
@brianlheim @jamshark70 updated the pull-request hopefully addressing all the review comments.

So following your feedback I've changed the recompile logic to use the in-memory version of the language config instead of the one in the file. If a changed is made in the config file via the IDE then the interpreter must be restarted to see it, I think this is reasonable, what do you think ?

miguel-negrao commented:
> I'm not confident enough in C++ to comment on the code, but:
> 
>     If a changed is made in the config file via the IDE then the interpreter must be restarted to see it, I think this is reasonable, what do you think ?
> 
> This is currently what happens when you change included or excluded paths using IDE preferences: a dialog pops up advising the user to reboot the interpreter.
> 
Yes, although with the exclude default paths switching to the lang file inthis PR, technically a recompile would be enough. But when using the IDE I don't see much difference between recompile and restart interpreter, so perhaps it's not worth adding the code to load the new lang file into memory after changing it in the IDE.

> We should make sure that the dialog pops up for this setting too.

That code is not touched, so it will continue to show that pop up.

miguel-negrao commented:
Also, before when changing the contents of the config file a recompile would show the new contents, with this PR a reboot of the interpreter will be needed.

jamshark70 commented:
> But when using the IDE I don't see much difference between recompile and restart interpreter...

Internally, there is a difference between recompiling within the same sclang process, or quitting sclang and launching a new process.

> That code is not touched, so it will continue to show that pop up.

Sure, but does the new setting trigger the pop-up? I would test it myself, but there's no way I will have time until the weekend, so I'm asking.

My main point is that the new setting should, with regard to IDE interface behavior and language reboot, behave exactly the same as the other language config settings:

- If done through the IDE (which directly manipulates yaml files), the user gets a pop-up advising to reboot the interpreter. The change doesn't take effect until the interpreter reboot rereads the config file.

- If done by an sclang method call, the change takes effect on the next recompile and is not persisted to disk automatically. (The user can do it by `write`.)

I think the "exclude" setting should not deviate from this, so that the documentation can explain one rule without too many exceptions. If there's a very good reason why it should be different, I'm open to it but it would have to be a strong case.

miguel-negrao commented:
> But when using the IDE I don't see much difference between recompile and restart interpreter...
> 
> Internally, there is a difference between recompiling within the same sclang process, or quitting sclang and launching a new process.

Yes, I'm aware of that. One difference is that it will be faster to recompile then to create a new process. I wonder, besides this difference is there any other difference for the user (other than the one introduced in this PR that the language config is persisted) ?
> 
>     That code is not touched, so it will continue to show that pop up.
> 
> Sure, but does the new setting trigger the pop-up? I would test it myself, but there's no way I will have time until the weekend, so I'm asking.

Ah, sorry, I didn't get that. Yes it does. The config is marked dirty when the toggle state is changed: 

https://github.com/miguel-negrao/supercollider/blob/0efe799f0e153e04c28bc1a4ab26b3a180750073/editors/sc-ide/widgets/settings/sclang_page.cpp#L66

> My main point is that the new setting should, with regard to IDE interface behavior and language reboot, behave exactly the same as the other language config settings:
> 
>     If done through the IDE (which directly manipulates yaml files), the user gets a pop-up advising to reboot the interpreter. The change doesn't take effect until the interpreter reboot rereads the config file.
> 
>     If done by an sclang method call, the change takes effect on the next recompile and is not persisted to disk automatically. (The user can do it by write.)
>

Yes, that is the way it is working in the current state of this PR.

jamshark70 commented:
> I wonder, besides this difference is there any other difference for the user (other than the one introduced in this PR that the language config is persisted) ?

Tail call optimization is, as I recall, not persisted anywhere (and there isn't really any need to). Turning it off for debugging requires a recompile, but a full restart wipes out the setting. So if there's no option to recompile in-process, there's no way to use that setting. (Note, nobody should be depending on TCO being off for their code to work. I don't think we need to add this to LanguageConfig... Nobody is asking to, I'm just heading off an argument before it starts :wink: )

> The config is marked dirty when the toggle state is changed:

Great, then it's handled correctly.

muellmusik commented:
What is the status of this PR?

miguel-negrao commented:
I haven't looked at this for a year and half, but looking at the comments I think the final version of the PR was never reviewed.

muellmusik commented:
Okay, in the first instance, it might be best if @brianlheim reviews the final version, since most of the changes are in response to his earlier comments. But I know he is busy with lots of SC work, so if someone else feels happy to dive in that would be welcome.

@miguel-negrao maybe in the meantime you could look at the conflicts and see if there is nothing serious with the current state of develop? Sorry I know that's a bother given this has been waiting, but you're probably the best person to judge. Most of them look minor.

brianlheim commented:
i won't review this until it is free from conflicts.

brianlheim commented:
sorry if that came off as a bit curt. to clarify, i don't review PRs with conflicts as a rule, because i believe it wouldn't be fair to other contributors. until that's fixed, i will prioritize PRs that can be merged.

miguel-negrao commented:
@muellmusik @brianlheim  I'll try to fix the conflicts soon.

geoffroymontel commented:
@miguel-negrao do you need help on fixing the conflicts for this PR ?

miguel-negrao commented:
Sorry I haven't done this yet. 
@geoffroymontel If you don't mind fixing the conflicts, go ahead. Otherwise I will try to do it soon
.

miguel-negrao commented:
Solving this conflict is slightly complicated because there have been whitespace changes meanwhile. Will have a look at the new whitespace guidelines, and will try to find a way to format my code changes in the same style to be able to see only the real conflicts.

geoffroymontel commented:
Sure ! It's due to the big C++ codebase reformatting.

I did the work in [this branch](https://github.com/geoffroymontel/supercollider/tree/miguel/topics/removeStandaloneSwitch-reformatted) and made a [PR to your repo](https://github.com/miguel-negrao/supercollider/pull/2).

But if you want to do the process again by yourself [here is the guide](https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions).

Let me know if I can help !

brianlheim commented:
@geoffroymontel thanks for doing that! you shouldn't merge that PR, but rather use it directly in a new PR and close this one.

miguel-negrao commented:
Ok, I think I successfully reformatted (thanks for the clang-format script !!) and rebased onto develop. Let's see if it clears the tests.

miguel-negrao commented:
I don't understand the error in travis. Is an formatting issue ? But it seems to have cleared it it on the other runs...

brianlheim commented:
@miguel-negrao there's a rare couple cases where long comments need multiple passes with clang-format to flow correctly. sorry about that! thanks for updating

miguel-negrao commented:
I think there was an issue with formatting. I've reformatted and hopefully fixed it.

miguel-negrao commented:
Ok, tests are passing. I haven't looked at this code in years, so I was going over it again, and it seems to make sense, and also I've built SC and quickly tested and seems to work, but I would feel better if someone else tested this also.

miguel-negrao commented:
In the original pull request message there isn't much about the motivation behind this change. Should I write a bit more about the reasoning here ? Would it be worthwhile to do an RFC, or is that not needed ? Let me know how to proceed.

elgiano commented:
IMO since this is a non-breaking new feature, there has been already a discussion, and you addressed all the relevant points, I don't think a PR is needed here! But please, anyone correct me if I'm wrong.

I looked through your code and tried it out, there are a few issues left:

- Library doesn't compile if 'excludeDefaultPath' is not present in YAML. Could be fixed like this:
```diff
--- a/lang/LangSource/SC_LanguageConfig.cpp
+++ b/lang/LangSource/SC_LanguageConfig.cpp
@@ -123,11 +123,12 @@ static void processBool(const char* nodeName, YAML::Node& doc, const std::functi
     if (item) {
         try {
             successFunc(item.as<bool>());
+            return;
         } catch (...) {
             postfl("Warning: Cannot parse config file entry \"%s\"\n", nodeName);
-            failFunc();
         }
-    }
+    };
+    failFunc(); // call failFunc if item is not found
 }
```
- popups for restarting sclang are not appearing in my scide
- when changing excludeDefaultPath from the IDE panel, no change is applied on recompile (but only on sclang process restart)
- sclang methods to get/set excludeDefaultPath? You did the primitive, but not the class methods.
- related help docs.

Other than these, it all looks good to me, and I see you addressed all discussed points.
About purpose: I tried to guess it from your other PRs that this is going in the direction of making SC projects possible, which is for sure an interesting feature. I agree that an explanation in this PR's description would make it easier to understand :)

elgiano commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -93,47 +103,51 @@ bool SC_LanguageConfig::removeIncludedDirectory(const Path& path) { return remov
+static void processBool(const char* nodeName, YAML::Node& doc, const std::function<void(bool)>& successFunc,
+                        const std::function<void(void)>& failFunc) {
+    const YAML::Node& item = doc[nodeName];
+    if (item) {
```
Library doesn't compile if 'excludeDefaultPaths' is not present in YAML. Could be fixed like this:
```diff
--- a/lang/LangSource/SC_LanguageConfig.cpp
+++ b/lang/LangSource/SC_LanguageConfig.cpp
@@ -123,11 +123,12 @@ static void processBool(const char* nodeName, YAML::Node& doc, const std::functi
     if (item) {
         try {
             successFunc(item.as<bool>());
+            return;
         } catch (...) {
             postfl("Warning: Cannot parse config file entry \"%s\"\n", nodeName);
-            failFunc();
         }
-    }
+    }
+    failFunc(); // call failFunc if item is not found
 }
```

elgiano commented on SCClassLibrary/Common/Core/LanguageConfig.sc:
```diff
@@ -43,4 +43,12 @@ LanguageConfig {
 		_LanguageConfig_setPostInlineWarnings
 		^this.primitiveFailed
 	}
+	*defaultPathsExcluded {
```
Wouldn't it be more consistent with the rest of the code, to call these `excludeDefaultPaths`? Everywhere else, e.g. postInlineComments, functions like these have the same name as in YAML

elgiano replied:
And, it would be nice to have these two functions documented LanguageConfig's help file as well

miguel-negrao commented:
@elgiano Thanks for the review. I will fix the issues you mentioned and add a new commit soon.

elgiano commented:
Thank you!

elgiano commented on testsuite/sclang/CMakeLists.txt:
```diff
@@ -3,5 +3,5 @@ file(GLOB tests *.scd)
     get_filename_component(test_name ${test} NAME_WE)
     set(test_target ${test_name}_run)
-    add_test(NAME ${test_target} COMMAND sclang -a -l ${CMAKE_BINARY_DIR}/build_sclang.cfg ${test})
+    add_test(NAME ${test_target} COMMAND sclang -l ${CMAKE_BINARY_DIR}/build_sclang.cfg ${test})
```
I grepped for "build_sclang.cfg" and another place showed up:
https://github.com/supercollider/supercollider/blob/00476b2a955402d9beff198c65faee5654a154fb/CMakeLists.txt#L574
Any reason why you wouldn't do that there too?

miguel-negrao replied:
No. Will change here also.

elgiano commented:
@miguel-negrao thanks! only one minor thing left (see above comment). It all works, I don't really know why one travis build is failing a test (maybe @brianlheim or @patrickdupuis know more about it?).

Last thing would be, and it's not a hard requirement, if you can squash all your commits into one. I tried it on my machine and it looks nice, clean and easier to follow :)

miguel-negrao commented:
> @miguel-negrao thanks! only one minor thing left (see above comment). It all works, I don't really know why one travis build is failing a test (maybe @brianlheim or @patrickdupuis know more about it?).

Yeah, the error seems related to math operators, could it be due to a time-out ?


> Last thing would be, and it's not a hard requirement, if you can squash all your commits into one. I tried it on my machine and it looks nice, clean and easier to follow :)

Sure thing, will squash. I myself prefer when it is a single commit.

elgiano commented:
> Yeah, the error seems related to math operators, could it be due to a time-out ?
Forgot to mention that I built your branch on my linux machine and it passes all tests (like it passes them on travis' xcode build)

miguel-negrao commented:
unrelated but, I had to change tools/CMakeLists.txt in order to use clang-format on my machine with debian 10 because the executable is named clang-format-diff-8 and not clang-format-diff.

miguel-negrao commented:
Also, regarding the reasoning for this PR, some language config files require the default include paths to be active while others require the default include paths not to be active, in order for the language to compile successfully, therefore it makes sense that that choice is made in the yaml file itself, so that it will work automatically.

elgiano commented:
> unrelated but, I had to change tools/CMakeLists.txt in order to use clang-format on my machine with debian 10 because the executable is named clang-format-diff-8 and not clang-format-diff.

And this change is (of course) not part of the present PR

miguel-negrao commented:
> > unrelated but, I had to change tools/CMakeLists.txt in order to use clang-format on my machine with debian 10 because the executable is named clang-format-diff-8 and not clang-format-diff.
> 
> And this change is (of course) not part of the present PR

No, indeed it isn't. I mentioned that only to ask if others had that issue, and what should one do about it ? Best solution is to symlink clang-format-diff-8 to clang-format-diff ?

elgiano commented:
I don't know enough about it, but that's what I would do. Additionally, if
you have different versions of clang installed, you could make a script to
change what that symlink is pointing to.

On Thu, Jun 11, 2020 at 5:58 PM Miguel Negrão <notifications@github.com>
wrote:

> unrelated but, I had to change tools/CMakeLists.txt in order to use
> clang-format on my machine with debian 10 because the executable is named
> clang-format-diff-8 and not clang-format-diff.
>
> And this change is (of course) not part of the present PR
>
> No, indeed it isn't. I mentioned that only to ask if others had that
> issue, and what should one do about it ? Best solution is to symlink
> clang-format-diff-8 to clang-format-diff ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/3733#issuecomment-642768789>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACNMWSA4QETLOILWQE4SGTRWD5LBANCNFSM4FAPICPA>
> .
>

brianlheim commented:
@miguel-negrao you can use the env vars SC_CLANG_FORMAT and SC_CLANG_FORMAT_DIFF

brianlheim commented:
> with this PR:
>    sclang -a option does nothing, only warns that language file should be used.

i don't think this description from the top comment is accurate anymore:

```
[bheim@mandible build]$ lang/sclang -a
compiling class library (debug build)...
	Found 866 primitives.
ERROR: There is a discrepancy.
numClassDeps 0   gNumClasses 82
sc3> 
```

can you please update the top comment to explain what this PR is doing? and if you have time, rebase it on latest develop (otherwise there are build errors with Qt 5.15 on Arch Linux)? then i am ready to review it.

miguel-negrao commented:
@brianlheim will update top post and rebase on latest develop.

miguel-negrao commented:
> @brianlheim will update top post and rebase on latest develop.

done

brianlheim commented:
thanks! sorry i didn't have time to get to this until now, i have been really busy recently. i am looking at it tonight. the new description is very helpful, thank you.

brianlheim commented:
just want to note that i've observed the following behavior --

- `-a` option overrides `excludeDefaultPaths: false` in the config file (this is what i would prefer)
- on a recompile with `-a` and `excludeDefaultPaths: false`, that overridden setting is preserved (this is what i would prefer)

brianlheim commented on HelpSource/Classes/LanguageConfig.schelp:
```diff
@@ -52,6 +52,9 @@ Remove path from class library exclude paths.
 Return the current config file path.
 
+METHOD:: excludeDefaultPaths
+Get or set whether default class library paths are included.
```
it would be good to note that:
- this value can be overridden if sclang is started with `-a`
- the default class library paths are `Platform.classLibraryDir`, and the system and user extensions directories

miguel-negrao replied:
Added note to the help file.

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -93,47 +103,53 @@ bool SC_LanguageConfig::removeIncludedDirectory(const Path& path) { return remov
+            failFunc();
+        }
+    }
+    failFunc(); // call failFunc if item is not found
```
this should be placed in the `else` branch otherwise it will be called twice if line 125 throws. you can also remove the `return` in the `try` block then.

brianlheim replied:
just saw @elgiano 's comment above. you didn't quite use the diff he suggested, which was to delete line 129 entirely. you could also do that. overall, i think the control flow is clearer with my suggestion, but it's up to you.

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -93,47 +103,53 @@ bool SC_LanguageConfig::removeIncludedDirectory(const Path& path) { return remov
+}
+
+static void processBool(const char* nodeName, YAML::Node& doc, const std::function<void(bool)>& successFunc,
+                        const std::function<void(void)>& failFunc) {
```
write this as `std::function<void()>`, the inner void is redundant

edit - did not intend for that to sound so philosophical haha

brianlheim commented on lang/LangSource/SC_LanguageConfig.cpp:
```diff
@@ -93,47 +103,53 @@ bool SC_LanguageConfig::removeIncludedDirectory(const Path& path) { return remov
+            successFunc(item.as<bool>());
+            return;
+        } catch (...) {
+            postfl("Warning: Cannot parse config file entry \"%s\"\n", nodeName);
```
would suggest changing this back to "WARNING", otherwise it is easy to miss in the startup messaging

miguel-negrao commented:
> thanks! sorry i didn't have time to get to this until now, i have been really busy recently. i am looking at it tonight. the new description is very helpful, thank you.

No worries, these are strange times with lots going on. Thank you for reviewing this !!

I've added the requested changes and rebased on develop again. Btw, should the explanation on the top post be added to the commit message or is it enough to have it here on the pull request ?

elgiano commented:
I think the only critical point about your commit message is that it says "change from command line parameter to ...", while finally this PR is adding a language file config option. That could be misleading and require lots of explanations (I imagine someone will have to make a changelog at some point), while if it was clearer that this is adding a feature without changing others, then I think it would be ok to leave all further explanations and motivations in the PR text, for the ones who want to dig deeper.

miguel-negrao commented:
Changed commit message to better reflect changes.

miguel-negrao commented:
Thank you @brianlheim and @elgiano for the multiple reviews ! Thanks @jamshark70 for the insightful comments and suggestions !
